src/midi/midi.cpp

```diff
@@ -168,6 +217,85 @@ uint8_t get_midi_channel(const uint8_t channel_status)
 	return channel_status & 0x0f;
 }
 
+static void output_note_off_for_active_notes(const uint8_t channel)
+{
+	assert(channel <= LastMidiChannel);
+
+	constexpr auto note_off_velocity = 64;
+	constexpr auto note_off_msg_len  = 3;
+	uint8_t buf[note_off_msg_len]    = {};
+
+	buf[0] = MidiStatus::NoteOff | channel;
+	buf[2] = note_off_velocity;
+
+	for (auto note = FirstMidiNote; note <= LastMidiNote; ++note) {
+		if (midi_state.IsNoteActive(channel, note)) {
+			buf[1] = note;
+
+			if (CaptureState & CAPTURE_MIDI) {
+				constexpr auto is_sysex = false;
+				CAPTURE_AddMidi(is_sysex, note_off_msg_len, buf);
+			}
+			midi.handler->PlayMsg(buf);
+		}
+	}
+}
+
+// Many MIDI drivers used in games send the "All Notes Off" Channel Mode
+// Message to turn off all active notes when switching between songs, instead
+// of properly sending Note Off messages for each individual note as required
+// by the MIDI specification (all Note On messages *must* be always paired
+// with Note Offs; the "All Notes Off" message must not be used as a shortcut
+// for that). E.g. all Sierra drivers exhibit this incorrect behaviour, while
+// LucasArts games are doing the correct thing and pair all Note On messages
+// with Note Offs.
+//
+// This hack can lead to "infinite notes" (hanging notes) when recording the
+// MIDI output into a MIDI sequencer, or when using DOSBox's raw MIDI output
+// capture functionality. What's worse, it can also result in multiple Note On
+// messages for the same note on the same channel in the recorded MIDI stream,
+// followed by a single Note Off only. While playing back the raw MIDI stream
+// is interpreted "correctly" on MIDI modules typically used in the 1990s,
+// it's up to the individual MIDI sequencer how to resolve this situation when
+// dealing with recorded MIDI data. This can lead lead to missing notes, and
+// it makes editing long MIDI recordings containing multiple songs very
+// difficult and error-prone.
+//
+// See page 20, 24, 25 and A-4 of the "The Complete MIDI 1.0 Detailed
+// Specification" document version 96.1, third edition (1996, MIDI
+// Manufacturers Association) for further details
+//
+// https://archive.org/details/Complete_MIDI_1.0_Detailed_Specification_96-1-3/
+//
```

Comment:
Extremely helpful discussion of the issue and the use of the new sanitize feature!

Reply:
Was afraid if it might be a bit too much info for some, but then I started thinking it would do the project a lot of good if literally 50% of the total LOC was helpful documentation like this, explaining varius bits and pieces about the hardware, standards, etc. Glad you appreciate it! 🎉